### ISSUE TITLE
Chemcam simualtion model

### ISSUE THREAD

**Issue by menelaos94** [menelaos94/Chemcam-Models#1]
This is a model endeavouring to simulate what was done in the predictive analyses of Chemcam data. 
%this script is function for which the inputs are T, Y i and q

%T is the training set of spectral data for number of distinct spectra. In
%our case, there were 263 corresponding to 66 distinct samples from
%ChemCam's data- 3-4 spectra for each sample which we averaged to get 66
%spectra.

%Each spectrum has a corresponding elemental composition which was obtained
%by evaluating key peaks using Laser Induced Breakdown Spectroscopy (LIBS)
%thus generating the matrix of compositions Y

% i is the sample number which we take out as a validation set, to
% validated the model which we build from the remaining 65 spectra

%q is the number of features (independent variables) which we would like to
%reduce to using Independent Component Analysis (ICA)

%We first run and ICA on our spectral data and reduce it to q features
%we then compute the Euclidean distance between the full spectra of our
%validation sample and each sample in the training set. This gives us an
%indication as to how "close" our samples are to one another.

%we proceed to sorting the norms, and then creating a minimum training set
%for our model. In this case we used 11 samples as a minimum. We compare
%the accuracy of the model with 11 samples to a model with 12 samples. And
%add samples as needed to obtain an optimal result (minimum error).

function [output]=nasatrial(T,Y,i,q)
    %i is the sample with we are validating of our sample
    XTRAIN=[T;Y];
    VALNEW=XTRAIN(:,i);
    XTRAIN(:,i)=[];
    Mdl=rica((T)',q);%icp model
    z = transform(Mdl,(T)');% icp result
    train_new_X=[z';Y];%reduced training set
    val_new=train_new_X(:,i);%reduced validation vector
    train_new_X(:,i)=[];%reduced training set

    
    [r, c]=size(train_new_X);

    diff=zeros(r,c-1);
    normal=zeros(1,65);
    for i=1:65
        diff(:,i)=val_new(:,1)-train_new_X(:,i);
        DIFFERENCE=diff;
        normal(1,i)=norm(DIFFERENCE(1:10,i));
    end
    
   
    
    [~,I]=sort(normal);
 
    i=11;
        while i<64
       

            trainfull=XTRAIN(:,I(1,1:i));
            trainfull_2=XTRAIN(:,I(1,1:i+1));
            trainfull=trainfull'; %resulting training set
            trainfull_2=trainfull_2'; %resulting training set
            XTRAIN_1=trainfull(:,1:end-10); %resulting xtraining set
            YTRAIN_1=trainfull(:,end-9:end);%resulting ytraining set

            XTRAIN_2=trainfull_2(:,1:end-10); %resulting xtraining set
            YTRAIN_2=trainfull_2(:,end-9:end);%resulting ytraining set
            
            YTRAIN_1(34,2)=0;%accounting for missing data
            YTRAIN_2(34,2)=0;%accounting for missing data
            YTRAIN_1(46,2)=0;%accounting for missing data
            YTRAIN_2(46,2)=0;%accounting for missing data
            YTRAIN_1(2,20)=0;%accounting for missing data
            YTRAIN_2(2,20)=0;%accounting for missing data
            
            YTRAIN_1(i+1:end,:)=[];%accounting for dimensionality
            YTRAIN_1(:,i+1:end)=[];%accounting for dimensionality
            YTRAIN_2(i+2:end,:)=[];%accounting for dimensionality
            YTRAIN_2(:,i+2:end)=[];%accounting for dimensionality
            
            
            [BETA1]=plsregress(XTRAIN_1,YTRAIN_1,10);%regress1
            [BETA2]=plsregress(XTRAIN_2,YTRAIN_2,10);%regress2
            YPRED_1=zeros(1,10);%predicted output for higher number of smaples
            for k=1:10
                YPRED_1(1,k)=dot((BETA1(:,k)),VALNEW(1:6144,1));%prediction
            end
            
            YPRED_2=zeros(1,10);%predicted output for lower number of smaples
            
            for m=1:10
                YPRED_2(1,m)=dot((BETA2(:,m)),VALNEW(1:6144,1));
            end

            if rmsee((YPRED_1),(VALNEW(11:20)))<rmsee((YPRED_2),(VALNEW(11:20))) 
                i=i+1;%evaluate error of models, and increase number of samples included to improve 
       
                output=YPRED_1;
            else
                output=YPRED_1;
                i=0;
            end
        end
    end
